### ISSUE TITLE
New "check-in policies page" gives incorrect directions 

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#930]
The following page:

https://github.com/trilinos/Trilinos/wiki/Policies-|-Safe-Checkin-Testing

gives incorrect directions for running the new checkin-test-sems.sh script.  It also does not mention anything about the `TRILINOS_DIR` environment variable.

@trilinos/framework

**Comment by mhoemmen** [comment]
Also, you need to `source` the script if you want the modules to stay in your path, e.g., for running `make`.

**Comment by bartlettroscoe** [comment]
>  It also does not mention anything about the TRILINOS_DIR environment variable.

If you follow the instructions exactly as described on that page (and symlink the checkin-test-sems.sh script), then you never need to set the `TRILINOS_DIR` env var.  It is set automatically inside of the script.  However, if you are on Mac OSX, that does not work and it gives you an error message.    In an attempt to improve on this error message, I just pushed the commit 7c9ad6d7088 which prints out:

  echo "ERROR: Cannot determine TRILINOS_DIR (you must be on a non-Linux system or you must have copied the script instead of symlinking it as per instructions). If you want to try to use this script on this system then please use standard directory structure or set TRILINOS_DIR manually in the env!  But this script and build process is currently only set up to support RHEL Linux 6 machines using the SEMS env."

Is that helpful?

@trilinos/muelu developers,

For what I can see, only one of two things is causing this:

1) You copied the the script checkin-test-sems.sh instead of symlinking it as instructed
2) You tired to use this on Mac OSX with a non-standard directory structure

Can someone please describe exactly what they are doing so that I can understand why you this is not working for you?

> also, you need to `source` the script if you want the modules to stay in your path, e.g., for running `make`.

That is true, but that use case is explained on this page with:

```
$ cd Trilinos/CHECKIN/
$ source ../cmake/load_ci_sems_dev_env.sh
$ cd MPI_RELEASE_DEBUG_SHARED_PT/
$ ./do-configure               # Reproduce configure failure
$ make -j<N>                   # Reproduce build failure
$ ctest -j<N> -R <test-name>   # Reproduce failing test(s)
```

That seems less mysterious to me but I will not fault people for wanting to just source checkin-test-sems.sh and be done with it.  To each their own.

On a RHEL 6 machine with the SEMS env, using checkin-test-sem.sh should be like falling off a log.  If it is not, then lets fix it so that it is.

**Comment by tawiesn** [comment]
I have a Sandia RHEL 7 machine with NO sems environment. Since SEMS modules were not available when my machine was set up, i've compiled my own development environment and set up my configurations for the (old) classic checkin script and i'm quite happy with it.
At the moment i have no time to set up and test the SEMS modules for RHEL 7. The procedure described by @mhoemmen in #924 works for me and allows me to locally test my changes in all my test configurations.

**Comment by bartlettroscoe** [comment]
> At the moment i have no time to set up and test the SEMS modules for RHEL 7

When you have some time, it should take just a few minutes to NFS the SEMS modules on your machine as explained here:

* https://sems.sandia.gov/content/sonsrn-nfs-tpl-server-mount-linux

If that process is not super easy, then we need to give feedback to the SEMS team so that they can improve it.

> The procedure described by @mhoemmen in #924 works for me and allows me to locally test my changes in all my test configurations.

That is good.  Sorry if that was not more obvious.  The only things that changed were the --default-builds and I removed the symlink to checkin-test.py.  But if you have your own local builds, then you were not likely using --default-builds anyway.  And to adjust to the removal of the symlink should be a simple one-line change.

Are there other developers who are also having trouble?

Specifically, what is "incorrect" about the instructions at:

* https://github.com/trilinos/Trilinos/wiki/Policies-%7C-Safe-Checkin-Testing

If  one follows the instructions exactly as listed, then you never even know about the existence of a `TRILINOS_DIR` env var.  And now with the updated error message, it should hopefully be more clear what the problem might be.

Can we close this, or is there something else that needs to be addressed?

**Comment by mhoemmen** [comment]
If it's possible to run the check-in test script without SEMS modules, then we can close this issue.  Otherwise, I would prefer to leave it up.  Backwards compatibility suggests that it would have been nice to support use of the old check-in test script for a little while at least.

**Comment by bartlettroscoe** [comment]
> If it's possible to run the check-in test script without SEMS modules, then we can close this issue.

Yes, you should be able to run the raw checkin-test.py script without the SEMS modules.  You just will not be able to run the --default-builds unless you have specify in your MPI_RELEASE_DEBUG_SHARED_PT.config file how to find everything needed for the new PT build of Trilinos (i.e. compilers, MPI, and all the TPLs starting with BLAS and ending with Netcdf).  I designed the file Trilinos/project-checkin-test-config.py to allow that if necessary.  But if you just pass in `--default-builds=` in your wrapper script, then you can define any builds you want and run them with `--st-extra-builds=<build0>,<build1>,...`.  Is that not working?

I can demonstrate this by setting up a checkin-test.py wrapper that uses the CASL VERA env to define the PT build.  The CASL VERA env has all of the TPLs required by the new PT build.  And that you can build all of that on your own machine if you want.

What is the use case that is not working?  Can someone please be specific so that I can help?

**Comment by mhoemmen** [comment]
The specific use case is that @trilinos/muelu developers had old check-in test scripts, and they didn't understand the directions for converting them to use the new system.  They asked me directly, and I explained to them over e-mail.  When I mentioned to @maherou that I had done so, he thought it would be a good idea for me to post this issue to illustrate the work-around.

**Comment by bartlettroscoe** [comment]
> The specific use case is that @trilinos/muelu developers had old check-in test scripts, and they didn't understand the directions for converting them to use the new system.

What specific instructions need to be added to:

* https://github.com/trilinos/Trilinos/wiki/Policies-%7C-Safe-Checkin-Testing

?

Can someone just please add them?

I can put back the base-level symlink Trilinos/checkin-test.py if that would help.

**Comment by bartlettroscoe** [comment]
Okay, I added back the base-level `checkin-test.py` symlink in commit 1c4e2a0 and I added the following note to https://github.com/trilinos/Trilinos/wiki/Policies-%7C-Safe-Checkin-Testing:

* The raw script `Trilinos/checkin-test.py` can still be used for doing local development work and testing or any workflow and can even be used for pushing (but please prefer to use the `checkin-test-sems.sh` script on a RHEL 6 machine with the SEMS env when pushing to the main Trilinos 'develop' branch).  However, unless you can define the `MPI_RELEASE_DEBUG_SHARED_PT.config` file to support the full set of PT packages and TPLs, then just disable that build by running with your own defined builds `--default-builds= --st-extra-builds=<build0>,<build1>,...`.

Does this address the problem raised in this issue?

**Comment by tawiesn** [comment]
>The raw script Trilinos/checkin-test.py can still be used for doing local development work and testing or any workflow and can even be used for pushing (but please prefer to use the checkin-test-sems.sh script on a RHEL 6 machine with the SEMS env when pushing to the main Trilinos 'develop' branch).

Thanks for the clarification. So, the idea is that everybody is supposed to push from a RHEL 6 (or RHEL 7) machine with SEMS using checkin-test-sems.sh. Using the old checkin-test script is still allowed to push changes (at least for a while).

**Comment by bartlettroscoe** [comment]
> So, the idea is that everybody is supposed to push from a RHEL 6 (or RHEL 7) machine with SEMS using checkin-test-sems.sh. Using the old checkin-test script is still allowed to push changes (at least for a while). At least until there is the new pull-request based system working which replaces all checkin scripts, i guess?

Yes, all of that is correct.

It is not required to run the checkin-test-sem.sh script in order to push to the 'develop' branch as I stated in my email from last week that I archived in https://github.com/trilinos/Trilinos/issues/482#issuecomment-266124179.  It is just that your best chance at not breaking the standard CI build is is to use `./checkin-test-sems.sh --do-all --push`.  After that, the next best thing is to use checkin-test.py script at all (and at least that marks commits that were tested for robust git bisection).  And for there it is a free-for-all (or cross fingers and push).

Is this good enough?
